File: examples/enterprise/python/ent_08_multi_source_join.py
"""Enterprise 08 · Multi-source join: Kafka events + Parquet lookup → enriched output

Stream-table join using pandas merge:
  1. Product catalog loaded from in-memory Parquet (the "table" side)
  2. Order events collected from Kafka (the "stream" side)
  3. pandas merge on product_id → enriched order with product metadata
  4. Enriched output written to Parquet

This mirrors krishiv.stream_table_join() in the Python API.

Prerequisites:
    make infra-up seed-kafka

Run:
    KAFKA_BOOTSTRAP=localhost:9092 python python/ent_08_multi_source_join.py
"""

import json
import os
import time

import pyarrow as pa
import pyarrow.parquet as pq
import pandas as pd
from confluent_kafka import Consumer

BOOTSTRAP = os.environ.get("KAFKA_BOOTSTRAP", "localhost:9092")
OUT_PATH = "/tmp/krishiv-enterprise-py-08-enriched.parquet"

CATALOG = pd.DataFrame([
    {"product_id": 1, "name": "Laptop Pro",  "category": "electronics", "unit_price": 1299.99},
    {"product_id": 2, "name": "Mouse",       "category": "electronics", "unit_price":   29.99},
    {"product_id": 3, "name": "Desk Chair",  "category": "furniture",   "unit_price":  349.99},
    {"product_id": 4, "name": "Monitor 4K",  "category": "electronics", "unit_price":  499.99},
    {"product_id": 5, "name": "USB Hub",     "category": "electronics", "unit_price":   39.99},
])

SYNTHETIC_EVENTS = [
    {"order_id": 1001, "product_id": 1, "customer": "alice", "qty": 1},
    {"order_id": 1002, "product_id": 2, "customer": "bob",   "qty": 3},
    {"order_id": 1003, "product_id": 3, "customer": "carol", "qty": 1},
    {"order_id": 1004, "product_id": 1, "customer": "dave",  "qty": 2},
    {"order_id": 1005, "product_id": 5, "customer": "eve",   "qty": 5},
    {"order_id": 1006, "product_id": 4, "customer": "frank", "qty": 1},
]


def main() -> None:
    print("=== Enterprise 08 (Python): Kafka + Parquet → Enriched output ===")

    events = _collect_kafka(timeout_s=12)
    if not events:
        print("  no live Kafka — using synthetic order events")
        events = SYNTHETIC_EVENTS

    events_df = pd.DataFrame(events)
    print(f"\n  {len(events_df)} stream events  ×  {len(CATALOG)} catalog rows")

    # Stream-table join.
    enriched = events_df.merge(
        CATALOG.rename(columns={"name": "product_name"}),
        on="product_id",
        how="left",
    )
    enriched["line_total"] = (enriched["qty"] * enriched["unit_price"]).round(2)
    enriched = enriched.sort_values("line_total", ascending=False)

    print("\n--- Enriched orders ---")
    cols = ["order_id", "customer", "product_name", "category", "unit_price", "qty", "line_total"]
    print(enriched[cols].to_string(index=False))

    # Category revenue summary.
    summary = enriched.groupby("category").agg(
        orders=("order_id", "count"),
        revenue=("line_total", "sum"),
    ).reset_index()
    print("\n--- Revenue by category ---")
    print(summary.to_string(index=False))

    # Persist.
    table = pa.Table.from_pandas(enriched)
    pq.write_table(table, OUT_PATH)
    print(f"\n✓ {len(enriched)} enriched rows written to {OUT_PATH}")


def _collect_kafka(timeout_s: float) -> list[dict]:
    consumer = Consumer({
        "bootstrap.servers": BOOTSTRAP,
        "group.id": "krishiv-ent-py-08",
        "enable.auto.commit": True,
        "auto.offset.reset": "earliest",
    })
    consumer.subscribe(["orders"])
    events: list[dict] = []
    deadline = time.monotonic() + timeout_s
    try:
        while time.monotonic() < deadline:
            msg = consumer.poll(timeout=1.0)
            if msg is None:
                continue
            if msg.error():
                break
            try:
                events.append(json.loads(msg.value().decode("utf-8")))
            except Exception:
                continue
    finally:
        consumer.close()
    return events


if __name__ == "__main__":
    main()

File: examples/enterprise/rust/src/bin/ent_08_multi_source_join.rs
//! Enterprise 08 · Multi-source join: Kafka events + Parquet lookup — embedded mode
//!
//! Demonstrates stream-table join via Krishiv SQL:
//!   1. Load a static product catalog from Parquet (the "table" side)
//!   2. Simulate order events from an InMemoryKafkaSource (the "stream" side)
//!   3. Collect both sides into the Session, run a JOIN query
//!   4. Write enriched output to Parquet
//!
//! Run:
//!   cargo run -p krishiv-enterprise-examples --bin ent_08_multi_source_join

use std::sync::Arc;

use anyhow::{Context, Result};
use arrow::array::{Float64Array, Int64Array, StringArray};
use arrow::datatypes::{DataType, Field, Schema};
use arrow::record_batch::RecordBatch;
use krishiv::Session;
use krishiv_connectors::kafka::InMemoryKafkaSource;
use krishiv_connectors::Source;
use parquet::arrow::ArrowWriter;
use tempfile::tempdir;

#[tokio::main]
async fn main() -> Result<()> {
    println!("=== Enterprise 08: Multi-source join (embedded) ===");

    let dir = tempdir()?;

    // ── Product catalog (Parquet — table side) ─────────────────────────────
    let catalog_path = dir.path().join("catalog.parquet");
    {
        let schema = Arc::new(Schema::new(vec![
            Field::new("product_id",  DataType::Int64,   false),
            Field::new("name",        DataType::Utf8,    false),
            Field::new("category",    DataType::Utf8,    false),
            Field::new("unit_price",  DataType::Float64, false),
        ]));
        let batch = RecordBatch::try_new(schema.clone(), vec![
            Arc::new(Int64Array::from(vec![1, 2, 3, 4, 5])),
            Arc::new(StringArray::from(vec!["Laptop Pro","Mouse","Desk Chair","Monitor 4K","USB Hub"])),
            Arc::new(StringArray::from(vec!["electronics","electronics","furniture","electronics","electronics"])),
            Arc::new(Float64Array::from(vec![1299.99, 29.99, 349.99, 499.99, 39.99])),
        ])?;
        let file = std::fs::File::create(&catalog_path)?;
        let mut w = ArrowWriter::try_new(file, schema, None)?;
        w.write(&batch)?;
        w.close()?;
    }

    // ── Order events (InMemoryKafkaSource — stream side) ───────────────────
    let event_schema = Arc::new(Schema::new(vec![
        Field::new("order_id",   DataType::Int64, false),
        Field::new("product_id", DataType::Int64, false),
        Field::new("customer",   DataType::Utf8,  false),
        Field::new("qty",        DataType::Int64, false),
    ]));

    let event_batch = RecordBatch::try_new(event_schema, vec![
        Arc::new(Int64Array::from(vec![1001, 1002, 1003, 1004, 1005, 1006])),
        Arc::new(Int64Array::from(vec![1, 2, 3, 1, 5, 4])),
        Arc::new(StringArray::from(vec!["alice","bob","carol","dave","eve","frank"])),
        Arc::new(Int64Array::from(vec![1, 3, 1, 2, 5, 1])),
    ])?;

    let mut source = InMemoryKafkaSource::new("orders", 0, 0, vec![event_batch]);
    let mut event_batches: Vec<RecordBatch> = Vec::new();

    while let Some(batch) = source.read_batch().await.context("read_batch")? {
        println!("  read {} events from InMemoryKafkaSource", batch.num_rows());
        event_batches.push(batch);
    }

    // ── SQL JOIN ───────────────────────────────────────────────────────────
    let session = Session::builder().build()?;
    session.register_parquet("product_catalog", &catalog_path)?;
    session.register_record_batches("order_events", event_batches)?;

    let enriched_df = session.sql(
        "SELECT
             o.order_id,
             o.customer,
             p.name         AS product_name,
             p.category,
             p.unit_price,
             o.qty,
             CAST(o.qty AS DOUBLE) * p.unit_price AS line_total
         FROM order_events o
         JOIN product_catalog p ON o.product_id = p.product_id
         ORDER BY line_total DESC",
    )?;

    let result = enriched_df.collect()?;
    let batches = result.into_batches();
    let rows: usize = batches.iter().map(|b| b.num_rows()).sum();

    println!("\n--- Enriched orders ---");
    let view_session = Session::builder().build()?;
    view_session.register_record_batches("enriched", batches.clone())?;
    println!("{}", view_session.sql("SELECT * FROM enriched")?.collect()?.pretty()?);

    // ── Category revenue summary ───────────────────────────────────────────
    let summary_session = Session::builder().build()?;
    summary_session.register_record_batches("enriched", batches.clone())?;
    let summary = summary_session.sql(
        "SELECT category, COUNT(*) AS orders, ROUND(SUM(line_total), 2) AS revenue \
         FROM enriched GROUP BY category ORDER BY revenue DESC"
    )?;
    println!("\n--- Revenue by category ---");
    println!("{}", summary.collect()?.pretty()?);

    // ── Write enriched output to Parquet ──────────────────────────────────
    let out_path = dir.path().join("enriched_orders.parquet");
    if let Some(first) = batches.first() {
        let schema = first.schema();
        let file = std::fs::File::create(&out_path)?;
        let mut w = ArrowWriter::try_new(file, schema, None)?;
        for b in &batches { w.write(b)?; }
        w.close()?;
    }

    println!("\n✓ {} enriched rows written to {}", rows, out_path.display());

    Ok(())
}
